### PR TITLE
[CH 12832] Recommendations

### PR DESCRIPTION
**Pull request by jjl014**
#### Updates:
* Add support for retrieving Recommendations results + tests
* Add support for tracking Recommendation events + tests

### CODE REVIEW

File: AutocompleteClient/FW/Logic/Request/CIOTrackRecommendationResultsViewData.swift [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+    let resultPage: Int?
+    let resultCount: Int?
+    let sectionName: String?
+    let resultID: String?
+
+    func url(with baseURL: String) -> String {
+        return String(format: Constants.TrackRecommendationResultsView.format, baseURL)
+    }
+
+    init(podID: String, url: String, numResultsViewed: Int? = nil, resultPage: Int? = nil, resultCount: Int? = nil, sectionName: String? = nil, resultID: String? = nil) {
```
Comment (crgee1) [inline AutocompleteClient/FW/Logic/Request/CIOTrackRecommendationResultsViewData.swift:28]: might need to default url to "Not Available"

File: AutocompleteClient/FW/Logic/Worker/ConstructorIO.swift [hunk 2]
```diff
@@ -177,6 +189,45 @@ public class ConstructorIO: CIOSessionManagerDelegate {
...
+    /// - Parameters:
+    ///   - podID: Pod ID
+    ///   - url: The current page URL the user viewed the recommendations on
+    ///   - numResultsViewed: The count of results that is visible to the user
+    ///   - resultPage: The current page that recommedantion result is on
+    ///   - resultCount: The total number of recommendation results
+    ///   - sectionName The name of the autocomplete section the term came from
+    ///   - resultID: Identifier of result set
+    ///   - completionHandler: The callback to execute on completion.
+    public func trackRecommendationResultsView(podID: String, url: String, numResultsViewed: Int? = nil, resultPage: Int? = nil, resultCount: Int? = nil, resultPositionOnPage: Int? = nil, sectionName: String? = nil, resultID: String? = nil, completionHandler: TrackingCompletionHandler? = nil) {
```
Comment (crgee1) [inline AutocompleteClient/FW/Logic/Worker/ConstructorIO.swift:203]: for `trackBrowseResultsLoaded` we do not surface the url as an argument, maybe we should do that as well and make the default "Not Available"

**Review by crgee1 — CHANGES_REQUESTED** [review]
Looks good, awesome work. Left some comments around `url` and I think the readme needs to be updated as well

File: AutocompleteClient/FW/Logic/Request/CIOTrackRecommendationResultsViewData.swift (hunk 1 above)
Comment (jjl014) [reply AutocompleteClient/FW/Logic/Request/CIOTrackRecommendationResultsViewData.swift]: Updated! 👍

File: AutocompleteClient/FW/Logic/Worker/ConstructorIO.swift (hunk 2 above)
Comment (jjl014) [reply AutocompleteClient/FW/Logic/Worker/ConstructorIO.swift]: Updated! 👍

**Review by crgee1 — APPROVED** [review]
nicely done!

File: AutocompleteClient/FW/Logic/Result/CIOPod.swift [hunk 3]
```diff
@@ -0,0 +1,24 @@
+//
+//  CIOPod.swift
```
Comment (rubikzube) [inline AutocompleteClient/FW/Logic/Result/CIOPod.swift:2]: `RecommendationsPod`?

File: AutocompleteClientTests/FW/API/Parser/RecommendationsResponseParserTests.swift [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+            let response = try self.parser.parse(recommendationsResponseData: data)
+            let pod = response.pod
+
+            XCTAssertEqual(pod.id, "item_page_1", "Pod ID should match the JSON response")
+            XCTAssertEqual(pod.displayName, "You may also like", "Pod Display Name should match the JSON response")
+        } catch {
+            XCTFail("Parser should never throw an exception when a valid JSON string is passed.")
+        }
+    }
+}
```
Comment (rubikzube) [inline AutocompleteClientTests/FW/API/Parser/RecommendationsResponseParserTests.swift:73]: I can haz test for checking an actual result?

File: AutocompleteClientTests/Resources/TestResource.swift [hunk 5]
```diff
@@ -32,9 +32,13 @@ struct TestResource {
         static let multipleInvalidVariationsValidVariationCount = 1
 
         static let multipleVariationsWithCustomDataJSONFilename = "response_search_json_variation_custom_data"
+        
+        static let recommendationsJSONFilename = "response_recommendations"
+        static let numberOfResultsInRecommendationsResponse = 10
     }
 
     static func load(name: String, type: String = "json") -> Data {
+        print(name, type)
```
Comment (rubikzube) [inline AutocompleteClientTests/Resources/TestResource.swift:41]: ???

File: AutocompleteClientTests/Test Utils/XCTest/XCTest+TimeoutFail.swift [hunk 6]
```diff
@@ -23,6 +23,6 @@ extension XCTestCase {
     }
 
     func wait(for expectation: XCTestExpectation) {
-        self.wait(for: [expectation], timeout: 1.0)
+        self.wait(for: [expectation], timeout: 10.0)
```
Comment (rubikzube) [inline AutocompleteClientTests/Test Utils/XCTest/XCTest+TimeoutFail.swift:26]: ???

**Review by rubikzube — CHANGES_REQUESTED** [review]

**Comment by rubikzube** [comment]
Plus add 1 integration test per new tracking event, as well as one test that makes a Recommenations request and gets back a live result.

**Pull request merged**